pkg/controller/engine.go

```diff
@@ -0,0 +1,207 @@
+package controller
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Error strings
+const (
+	errCreateCache      = "cannot create new cache"
+	errCreateController = "cannot create new controller"
+	errCrashCache       = "cache error"
+	errCrashController  = "controller error"
+	errWatch            = "cannot setup watch"
+)
+
+// A NewCacheFn creates a new controller-runtime cache.
+type NewCacheFn func(cfg *rest.Config, o cache.Options) (cache.Cache, error)
+
+// A NewControllerFn creates a new controller-runtime controller.
+type NewControllerFn func(name string, m manager.Manager, o controller.Options) (controller.Controller, error)
+
+// The default new cache and new controller functions.
+var (
+	DefaultNewCacheFn      NewCacheFn      = cache.New
+	DefaultNewControllerFn NewControllerFn = controller.NewUnmanaged
+)
+
+// An Engine manages the lifecycles of controller-runtime controllers (and their
+// caches). The lifecycles of the controllers are not coupled to lifecycle of
+// the engine, nor to the lifecycle of the controller manager it uses.
+type Engine struct {
+	mgr manager.Manager
+
+	started map[string]chan struct{}
+	errors  map[string]error
+	mx      sync.RWMutex
+
+	newCache NewCacheFn
+	newCtrl  NewControllerFn
+}
+
+// An EngineOption configures an Engine.
+type EngineOption func(*Engine)
+
+// WithNewCacheFn may be used to configure a different cache implementation.
+// DefaultNewCacheFn is used by default.
+func WithNewCacheFn(fn NewCacheFn) EngineOption {
+	return func(e *Engine) {
+		e.newCache = fn
+	}
+}
+
+// WithNewControllerFn may be used to configure a different controller
+// implementation. DefaultNewControllerFn is used by default.
+func WithNewControllerFn(fn NewControllerFn) EngineOption {
+	return func(e *Engine) {
+		e.newCtrl = fn
+	}
+}
+
+// NewEngine produces a new Engine.
+func NewEngine(mgr manager.Manager, o ...EngineOption) *Engine {
+	e := &Engine{
+		mgr: mgr,
+
+		started: make(map[string]chan struct{}),
+		errors:  make(map[string]error),
+
+		newCache: DefaultNewCacheFn,
+		newCtrl:  DefaultNewControllerFn,
+	}
+
+	for _, eo := range o {
+		eo(e)
+	}
+
+	return e
+}
+
+// IsRunning indicates whether the named controller is running - i.e. whether it
+// has been started and does not appear to have crashed.
+func (e *Engine) IsRunning(name string) bool {
+	e.mx.RLock()
+	defer e.mx.RUnlock()
+
+	_, running := e.started[name]
+	return running
+}
+
+// Err returns any error encountered by the named controller. The returned error
+// is always nil if the named controller is running.
+func (e *Engine) Err(name string) error {
```

Comment:
Yeah, something like that. I figured it could be used to log or emit a condition if and when we discover that the controller has crashed. could be something like:

```go
if !ctrl.IsRunning(name) {
    if ctrl.Err(name) != nil {
       log.Debug("womp womp", "error", err)
    }
    _ = ctrl.Start(name, o)
}
```